### PR TITLE
Fix race condition surrounding DB reopening during fork

### PR DESCRIPTION
**Pull request by lucasz93**
- [X] Added clear title that can be used to generate release notes

Hi all!

I was consistently running into this race condition in my project. One thread locked a copy of SQLiteHandleCache::sMutex_ while another thread was forking. Then the atfork callback tried to lock sMutex_. This caused a deadlock because the thread that owned the mutex isn't duplicated in a fork.

Kind regards,
-Lucas

### CODE REVIEW

File: src/iso19111/factory.cpp [hunk 1]
```diff
@@ -627,8 +627,23 @@ SQLiteHandleCache::getHandle(const std::string &path, PJ_CONTEXT *ctx) {
...
-        pthread_atfork(nullptr, nullptr,
-                       []() { SQLiteHandleCache::get().invalidateHandles(); });
+        pthread_atfork(
+            []() {
+                // This mutex needs to be acquired by 'invalidateHandles()'.
+                // The forking thread needs to own this mutex during the fork.
+                // Otherwise there's an opporunity for another thread to own
+                // the mutex during the fork, leaving the child process unable
+                // to acquire the mutex in invalidateHandles().
+                SQLiteHandleCache::get().sMutex_.lock();
```
Comment (jjimenezshaw) [inline src/iso19111/factory.cpp:637]: Isn't this lock conflicting with the lock at the beginning of the function?

Comment (lucasz93) [reply src/iso19111/factory.cpp:637]: Not to my knowledge. These callbacks run during a call to fork() whereas the lock at the beginning of the function is protecting the installation of the callbacks.

**Comment by rouault** [comment]
@lucasz93 Are you sure that playing with std::mutex around fork() is safe ? Particularly unlocking in a child process a std::mutex locked by its parent. I'm afraid we are in unspecified territory when mixing C++ mutex and fork()

**Comment by rouault** [comment]
ok, let's merge this. We'll see if that cause issues

**Pull request merged**